- Added support for grouping, merging, and discarding events

vicng opened the pull request:
- Added BlockMoveEvent to simplify tracking an individual Block's movement changes
- Changed Workspace listener from willRemoveBlock: to didRemoveBlock:, as listeners needed the latter, not the former
- Changed behavior so calls to fire (and group) pending events are done at the View layer, not the Layout layer
- Added method to MutatorLayout to indicate when starting a mutator 'session'. This is needed when replaying events as old mutators were storing stale data from previous unrelated sessions.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-ios/329)
<!-- Reviewable:end -->

vicng commented:
+@corydcode +@roboerikg 

---

Review status: 0 of 39 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329#-:-KfUSM4LGV9l0WZxm6r3:b4ivu6b)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 0 of 39 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Control/EventManager.swift, line 90 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhRuJvPV8vrJeibe7C:-KfhRuJvPV8vrJeibe7D:bak6ty0) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Control/EventManager.swift#L90)):*
> ```Swift
>    */
>   public func firePendingEvents() {
>     guard !_firingEvents else {
> ```

If this happens then there are two threads firing events. You need a sync block to prevent setting _firePendingEventsAgain just after the other thread checks its value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329)*
<!-- Sent from Reviewable.io -->

CoryDCode commented:
Reviewed 24 of 39 files at r1.
Review status: 24 of 39 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Layout/MutatorLayout.swift, line 62 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhQf289YIK6oFO4d05:-KfhQf289YIK6oFO4d06:b4o7pqo) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Layout/MutatorLayout.swift#L62)):*
> ```Swift
>    behavior for their mutators.
>    */
>   open func beginMutationSession() {
> ```

Mutators don't have a default mutation UI, and I think the only place this should be getting called is in the mutator UI on a per-mutator basis, should this be an abstract function?

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1192 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhPHbNBZ8SPJ6E4msk:-KfhPHbNBZ8SPJ6E4msl:bfy71v2) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/UI/View Controllers/WorkbenchViewController.swift#L1192)):*
> ```Swift
>           let mutatorLayout = block.mutator?.layout
>           let xml = try AEXMLDocument(xml: value)
>           mutatorLayout?.beginMutationSession()
> ```

Why does this need to be here? Isn't a "mutation session" a client session-specific concept?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 24 of 39 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Control/EventManager.swift, line 90 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhRuJvPV8vrJeibe7C:-KfhWhJ_Cz0OarAUq9QK:b2e54ur) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Control/EventManager.swift#L90)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

If this happens then there are two threads firing events. You need a sync block to prevent setting _firePendingEventsAgain just after the other thread checks its value.
</blockquote></details>

Discussed offline. This is for events that trigger more pending events. Fine as is, but please update the comment to make it clear this is a recursive call that's being delayed.

---

*[Source/Control/EventManager.swift, line 115 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhX0aC7Eow8ejJO-24:-KfhX0aC7Eow8ejJO-25:bn6t9wn) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Control/EventManager.swift#L115)):*
> ```Swift
>       // Immediately fire the next batch of events
>       _firePendingEventsAgain = false
>       firePendingEvents()
> ```

Note: It's possible this recurses infinitely and crashes. That's probably desirable so devs find it quickly.

---

*[Source/Layout/MutatorLayout.swift, line 62 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhYfQbkhOjCiNExkqg:-KfhYfQbkhOjCiNExkqh:bld991r) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Layout/MutatorLayout.swift#L62)):*
> ```Swift
>    behavior for their mutators.
>    */
>   open func beginMutationSession() {
> ```

Why is this on the base class instead of part of the editor UI? You don't have a common mutator editor UI, so adding this here doesn't make it easier on developers to implement.

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1192 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhPHbNBZ8SPJ6E4msk:-KfhZlUz3B_3qYGToIRc:bf778r2) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/UI/View Controllers/WorkbenchViewController.swift#L1192)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

Why does this need to be here? Isn't a "mutation session" a client session-specific concept?
</blockquote></details>

Agreed. If the mutator needs to do extra work it should do so in performMutation, not have a separate method that must be called prior to applying a mutation.

---

*[Source/UI/Views/FieldColorView.swift, line 116 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-Kfhbnb18yIfMGO_KWfP:-Kfhbnb18yIfMGO_KWfQ:b-ps34zv) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/UI/Views/FieldColorView.swift#L116)):*
> ```Swift
> 
> extension FieldColorView: FieldColorPickerViewControllerDelegate {
>   public func fieldColorPickerViewController(
> ```

Side note: Isn't this the old naming convention?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329)*
<!-- Sent from Reviewable.io -->

vicng commented:
Review status: 19 of 39 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Control/EventManager.swift, line 90 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhRuJvPV8vrJeibe7C:-Kfi5Tf7oicjPg23_lom:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Control/EventManager.swift#L90)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Discussed offline. This is for events that trigger more pending events. Fine as is, but please update the comment to make it clear this is a recursive call that's being delayed.
</blockquote></details>

Done.

---

*[Source/Control/EventManager.swift, line 115 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhX0aC7Eow8ejJO-24:-Kfi47a4IwDvyz-Y_dBW:bwygxsx) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Control/EventManager.swift#L115)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Note: It's possible this recurses infinitely and crashes. That's probably desirable so devs find it quickly.
</blockquote></details>

Noted.

---

*[Source/Layout/MutatorLayout.swift, line 62 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhQf289YIK6oFO4d05:-Kfi3UKeqLBQeku14fn1:bvtwqpp) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Layout/MutatorLayout.swift#L62)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

Mutators don't have a default mutation UI, and I think the only place this should be getting called is in the mutator UI on a per-mutator basis, should this be an abstract function?
</blockquote></details>

Good point. Reverted back and figured out another way to solve my problem (which was old state being cached while replaying events).

---

*[Source/Layout/MutatorLayout.swift, line 62 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhYfQbkhOjCiNExkqg:-Kfi3SnxrH-Bw4g-WFE3:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/Layout/MutatorLayout.swift#L62)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Why is this on the base class instead of part of the editor UI? You don't have a common mutator editor UI, so adding this here doesn't make it easier on developers to implement.
</blockquote></details>

Done.

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1192 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-KfhPHbNBZ8SPJ6E4msk:-Kfi3hSZG9nCs-nVnEun:bt2vr1b) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/UI/View Controllers/WorkbenchViewController.swift#L1192)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Agreed. If the mutator needs to do extra work it should do so in performMutation, not have a separate method that must be called prior to applying a mutation.
</blockquote></details>

Fixed.

---

*[Source/UI/Views/FieldColorView.swift, line 116 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-Kfhbnb18yIfMGO_KWfP:-Kfi3mh1qKuauavUXnJz:b-ted8so) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/UI/Views/FieldColorView.swift#L116)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Side note: Isn't this the old naming convention?
</blockquote></details>

Which part specifically? I think it's correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 19 of 39 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/UI/Views/FieldColorView.swift, line 116 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-Kfhbnb18yIfMGO_KWfP:-Kfi78NvuQCTab5vdw5a:b-ieuovp) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/UI/Views/FieldColorView.swift#L116)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Which part specifically? I think it's correct.
</blockquote></details>

This is Swift 2 style where the method name is the object instead of the action. Isn't the standard in swift3 for this to be more like colorChanged( _ viewController, newColor)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Feel free to ignore the comment on naming if I'm mistaken.

---

Review status: 19 of 39 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329#-:-KfiDheFIPDUNNcJsZ5w:b-geohp5)*
<!-- Sent from Reviewable.io -->

vicng commented:
Review status: 19 of 39 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[Source/UI/Views/FieldColorView.swift, line 116 at r1](https://reviewable.io:443/reviews/google/blockly-ios/329#-Kfhbnb18yIfMGO_KWfP:-KfiEPdeaL8XdAa3WL1x:bgkevkf) ([raw file](https://github.com/google/blockly-ios/blob/90db47d78f7ec420c1bd69bc750626e4195ed8d2/Source/UI/Views/FieldColorView.swift#L116)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

This is Swift 2 style where the method name is the object instead of the action. Isn't the standard in swift3 for this to be more like colorChanged( _ viewController, newColor)
</blockquote></details>

As discussed offline, Apple hasn't given direction for naming conventions of callback methods. Maybe in Swift 4 :P

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/329)*
<!-- Sent from Reviewable.io -->
